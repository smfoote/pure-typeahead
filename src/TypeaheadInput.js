import React, {Component} from 'react'

const noop = () =>{};

export default class TypeaheadInput extends Component {
  static defaultProps = {
    onKeyDown: noop,
    onFocus: noop,
    onKeyUp: noop,
  }

  onChange(evt) {
    this.props.onChange(evt);
  }

  onKeyDown(evt) {
    const { key } = evt;
    switch(key) {
      case 'ArrowDown':
      case 'ArrowUp':
        evt.preventDefault();
        this.props.arrowKeyPressed(key);
        break;
      case 'Enter':
        this.props.enterKeyPressed(evt);
        break;
      case 'Escape':
        this.props.escapeKeyPressed();
        break;
    }
    this.props.onKeyDown(evt);
  }

  focus() {
    this.inputRef.focus();
  }

  render() {
    const { children } = this.props;
    return (
      <input
        ref={input => this.inputRef = input}
        id={this.props.id}
        className={this.props.className}
        type={this.props.type || 'text'}
        placeholder={this.props.placeholder}
        value={this.props.value}
        autoComplete="off"
        onChange={(evt) => this.onChange(evt)}
        onKeyDown={(evt) => this.onKeyDown(evt)}
        onKeyUp={this.props.onKeyUp}
        onFocus={this.props.onFocus}
      />)
  }
}
