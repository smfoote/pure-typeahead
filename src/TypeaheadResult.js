import React, {Component} from 'react'

export default class TypeaheadResult extends Component {
  render() {
    const { children } = this.props;
    return (
      <typeahead-result
        onClick={this.props.onSelect}
        class={this.props.isHighlighted ? 'highlighted' : null}
      >
        {children}
      </typeahead-result>
    );
  }
}
