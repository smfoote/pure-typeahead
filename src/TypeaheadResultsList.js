import React, {Component} from 'react'
import TypeaheadResult from './TypeaheadResult'

export default class TypeaheadResultsList extends Component {

  render() {
    let currentIndex = -1;
    const resultValues = [];
    const children = React.Children.map(this.props.children, child => {
      if (child && child.type === TypeaheadResult) {
        // scope the index to avoid closure problems
        const idx = currentIndex = currentIndex + 1;
        resultValues.push(child.props.value);
        return React.cloneElement(child, {
          isHighlighted: idx === this.props.highlightedIndex,
          _onSelect: (value) => { this.props._onSelect(value); },
        })
      }
      return child;
    });
    this.props.onResultsUpdate(resultValues);
    return <typeahead-results-list class={this.props.className}>
      {children}
    </typeahead-results-list>
  }
}
