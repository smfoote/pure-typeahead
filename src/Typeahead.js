import React, {Component} from 'react'

import TypeaheadResultsList from './TypeaheadResultsList';
import TypeaheadInput from './TypeaheadInput';

export default class Typeahead extends Component {
  static defaultProps = {
    onBlur: () => {},
    onDismiss: () => {}
  }

  constructor(props) {
    super(props);
    this.state = {
      highlightedIndex: -1
    };
  }

  navigateList = (dir) => {
    this.resultsList.navigateList(dir);
  }

  updateHighlightedIndex = (highlightedIndex) => {
    this.setState({ highlightedIndex });
  }

  enterKeyPressed = (evt) => {
    // If a result is highlighted, select it
    if (this.state.highlightedIndex !== -1) {
      evt.preventDefault();
      this.selectHighlightedResult();
      // Reset highlighted index
      this.setState({highlightedIndex: -1});
    }
  }

  selectHighlightedResult = () => {
    this.resultsList.selectResult(this.state.highlightedIndex);
  }

  dismissTypeahead = () => {
    this.props.onDismiss();
  }

  onBlur = (evt) => {
    const { relatedTarget } = evt;
    if (!relatedTarget || !this.element.contains(relatedTarget)) {
      this.props.onBlur(evt);
    }
  }

  render() {
    const children = React.Children.map(this.props.children, child => {
      if (child) {
        if (child.type === TypeaheadResultsList) {
          return React.cloneElement(child, {
            highlightedIndex: this.state.highlightedIndex,
            updateHighlightedIndex: this.updateHighlightedIndex,
            ref: (ref => this.resultsList = ref)
          });
        } else if (child.type === TypeaheadInput) {
          return React.cloneElement(child, {
            arrowKeyPressed: this.navigateList,
            enterKeyPressed: this.enterKeyPressed,
            escapeKeyPressed: this.dismissTypeahead
          });
        }
      }
      return child;
    });
    return <pure-typeahead
      class={this.props.className}
      onBlur={this.onBlur}
      ref={ref => this.element = ref}
    >{children}</pure-typeahead>
  }
}
