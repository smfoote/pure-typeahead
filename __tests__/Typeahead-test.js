import React from 'react';
import Enzyme, { mount, shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import Adapter from 'enzyme-adapter-react-16';

import Typeahead from '../src/Typeahead';
import TypeaheadInput from '../src/TypeaheadInput';
import TypeaheadResultsList from '../src/TypeaheadResultsList';
import TypeaheadResult from '../src/TypeaheadResult';

Enzyme.configure({ adapter: new Adapter() });

const RESULT_VALUE = { some: 'value' };
const shallowSetup = () => {
  const props = {
    onDismiss: jest.fn(),
    onSelect: jest.fn()
  };
  return {
    wrapper: shallow(<Typeahead {...props}/>),
    props
  };
};

const mountSetup = () => {
  const props = {
    onDismiss: jest.fn(),
    onSelect: jest.fn()
  };
  return {
    wrapper: mount(<Typeahead {...props}>
      hi
      <TypeaheadInput value="hello" onChange={jest.fn()}/>
      <TypeaheadResultsList>
        <TypeaheadResult value={RESULT_VALUE}>RESULTS</TypeaheadResult>
        <TypeaheadResult value={'hello'}>hello</TypeaheadResult>
        <TypeaheadResult value={'world'}>world</TypeaheadResult>
      </TypeaheadResultsList>
    </Typeahead>),
    props
  };
};

describe('Typeahead', () => {
  it('renders self and subcomponents', () => {
    const { wrapper } = shallowSetup();
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('renders the children', () => {
    const { wrapper } = mountSetup();
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('updates the highlighted index when an arrow key is pressed', () => {
    const { wrapper } = mountSetup();
    wrapper.setState({highlightedIndex: -1});
    wrapper.find('TypeaheadInput').props().arrowKeyPressed('ArrowDown');
    expect(wrapper.state().highlightedIndex).toBe(0);
  });

  it('should navigate the list, go around the horn', () => {
    const { wrapper } = mountSetup();
    wrapper.find('TypeaheadInput').props().arrowKeyPressed('ArrowDown');
    expect(wrapper.state().highlightedIndex).toBe(0);
    wrapper.find('TypeaheadInput').props().arrowKeyPressed('ArrowDown');
    expect(wrapper.state().highlightedIndex).toBe(1);
    wrapper.find('TypeaheadInput').props().arrowKeyPressed('ArrowDown');
    expect(wrapper.state().highlightedIndex).toBe(2);
    wrapper.find('TypeaheadInput').props().arrowKeyPressed('ArrowDown');
    expect(wrapper.state().highlightedIndex).toBe(0);
    wrapper.find('TypeaheadInput').props().arrowKeyPressed('ArrowUp');
    expect(wrapper.state().highlightedIndex).toBe(2);
    wrapper.find('TypeaheadInput').props().arrowKeyPressed('ArrowUp');
    expect(wrapper.state().highlightedIndex).toBe(1);
    wrapper.find('TypeaheadInput').props().arrowKeyPressed('ArrowUp');
    expect(wrapper.state().highlightedIndex).toBe(0);
    wrapper.find('TypeaheadInput').props().arrowKeyPressed('ArrowUp');
    expect(wrapper.state().highlightedIndex).toBe(2);
  });

  it('should call props.onDismiss when escapeKeyPressed is called', () => {
    const { wrapper, props } = mountSetup();
    wrapper.find('TypeaheadInput').props().escapeKeyPressed();
    expect(props.onDismiss).toHaveBeenCalled();
  });

  it('should set the highlightedIndex state on updateHighlightedIndex', () => {
    const { wrapper } = mountSetup();
    wrapper.find('TypeaheadResultsList').props().updateHighlightedIndex(1);
    expect(wrapper.state().highlightedIndex).toBe(1);
  });

  it('should select the highlighted result', () => {
    const { wrapper, props } = mountSetup();
    const highlightedIndex = 0;
    const evt = { preventDefault: jest.fn() };
    wrapper.setState({highlightedIndex});
    wrapper.find('TypeaheadInput').props().enterKeyPressed(evt);
    expect(props.onSelect).toHaveBeenCalledWith(RESULT_VALUE);
    expect(evt.preventDefault).toHaveBeenCalledTimes(1);
    expect(wrapper.state('highlightedIndex')).toBe(-1);
  });

  it('should not select a result or preventDefault if highlightedindex is -1 state\'s highlightedIndex', () => {
    const { wrapper } = mountSetup();
    const highlightedIndex = -1;
    const evt = { preventDefault: jest.fn() };
    wrapper.select = jest.fn();
    wrapper.setState({highlightedIndex});
    wrapper.find('TypeaheadInput').props().enterKeyPressed(evt);
    expect(wrapper.select).toHaveBeenCalledTimes(0);
    expect(evt.preventDefault).toHaveBeenCalledTimes(0);
    expect(wrapper.state('highlightedIndex')).toBe(-1);
  });
});
