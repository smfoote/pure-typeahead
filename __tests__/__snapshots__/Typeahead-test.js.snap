// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Typeahead renders self and subcomponents 1`] = `<pure-typeahead />`;

exports[`Typeahead renders the children 1`] = `
<Typeahead
  onDismiss={[Function]}
>
  <pure-typeahead>
    hi
    <TypeaheadInput
      arrowKeyPressed={[Function]}
      enterKeyPressed={[Function]}
      escapeKeyPressed={[Function]}
      key=".1"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      value="hello"
    >
      <input
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        type="text"
        value="hello"
      />
    </TypeaheadInput>
    <TypeaheadResultsList
      highlightedIndex={-1}
      key=".2"
      updateHighlightedIndex={[Function]}
    >
      <typeahead-results-list>
        RESULTS
      </typeahead-results-list>
    </TypeaheadResultsList>
  </pure-typeahead>
</Typeahead>
`;
